### PR TITLE
Add buffer and fixed # of workers to Pipeline

### PR DESCRIPTION
**Pull request by cstahl1**
Use a fixed sized queue and a fixed number of workers to ensure that the
memory footprint cannot blow up when the pipeline is overloaded. This
ensures that requests are responded to before the processing is done,
which allows for the pipeline to shed load as needed.

### CODE REVIEW

File: pkg/collector/pipeline.go [hunk 1]
```diff
@@ -69,23 +93,24 @@ func (p *Pipeline) AddProcessor(processor ReportProcessor) {
 
 // ProcessReports extracts reports from a POST upload payload, as defined by the
 // Reporting spec, and runs all of the processors in the pipeline against each
-// report.
-func (p *Pipeline) ProcessReports(ctx context.Context, w http.ResponseWriter, r *http.Request) *ReportBatch {
+// report. Returns true if the request was dropped due to a full queue and false
```
Comment (sburnett) [inline pkg/collector/pipeline.go]: This return semantics feel backwards. I'd expect this to return true if things succeed.

Also, since there seem to be multiple ways this can fail (not just if the queue is full) you might make this return an error instead.

File: pkg/collector/pipeline.go [hunk 2]
```diff
@@ -102,15 +127,27 @@ func (p *Pipeline) ProcessReports(ctx context.Context, w http.ResponseWriter, r
 	err = decoder.Decode(&reports.Reports)
 	if err != nil {
 		http.Error(w, err.Error(), http.StatusBadRequest)
-		return nil
+		return true
```
Comment (sburnett) [inline pkg/collector/pipeline.go]: Why does this case count as dropped while the cases above don't? Feels inconsistent to me.

File: pkg/collector/pipeline.go [hunk 3]
```diff
@@ -49,17 +49,41 @@ func (c nowClock) Now() time.Time {
 var defaultClock nowClock
 
 // Pipeline is a series of processors that should be applied to each report that
-// the collector receives.
+// the collector receives. It uses a fixed number of workers to process the reports
+// and a fixed sized queue that the workers read from. If the queue fills, reports
+// are dropped.
 type Pipeline struct {
```
Comment (sburnett) [inline pkg/collector/pipeline.go]: My Go might be a weak, but if you want to prevent someone from instantiating one of these directly (without calling NewPipeline, for example), could you call this "pipeline" instead of "Pipeline"?

File: pkg/pipelinetest/pipelinetest.go [hunk 4]
```diff
@@ -172,7 +176,10 @@ func (p *PipelineTest) Run(t *testing.T) {
 				}
 
 				var response httptest.ResponseRecorder
-				batch := p.Pipeline.ProcessReports(context.Background(), &response, request)
+				p.Pipeline.ProcessReports(context.Background(), &response, request)
+				log.Printf("jarls about to read from channel")
```
Comment (sburnett) [inline pkg/pipelinetest/pipelinetest.go]: Probably want to get rid of these :-)

File: pkg/collector/pipeline.go [hunk 5]
```diff
@@ -49,17 +49,41 @@ func (c nowClock) Now() time.Time {
...
+// NewTestPipelineWithBuffer creates a new Pipeline with a specified buffer size and clock.
+// This should only be used for testing.
+func NewTestPipelineWithBuffer(clock Clock, bufferSize int64) *Pipeline {
+	return setupPipeline(context.Background(), clock, bufferSize, defaultNumWorkers)
+}
+
+func setupPipeline(ctx context.Context, clock Clock, bufferSize int64, numWorkers int) *Pipeline {
+	p := &Pipeline{clock: clock, c: make(chan *ReportBatch, bufferSize)}
+	for i := 0; i < numWorkers; i++ {
+		go p.runPipeline(ctx)
```
Comment (sburnett) [inline pkg/collector/pipeline.go]: Are you planning to add some way of cancelling / cleaning these up in a follon PR? Seems important.

Comment (cstahl1) [reply pkg/collector/pipeline.go]: On the one hand, this is for something that should just run forever, so I'm not clear that it's super important, but I have added a Close method that does so, which will be helpful for tests

File: pkg/pipelinetest/pipelinetest.go (hunk 4 above)
Comment (cstahl1) [reply pkg/pipelinetest/pipelinetest.go]: :-)

File: pkg/collector/pipeline.go (hunk 3 above)
Comment (cstahl1) [reply pkg/collector/pipeline.go]: Maybe, but there's no need to do that. I added to the comment stating that the zero initialized version is not useful.

File: pkg/collector/pipeline.go (hunk 2 above)
Comment (cstahl1) [reply pkg/collector/pipeline.go]: That is a mistake and should be false.

File: pkg/collector/pipeline.go (hunk 1 above)
Comment (cstahl1) [reply pkg/collector/pipeline.go]: My concern is isolating drops due to a full queue as compared to every other type of error (e.g. bad request, not a POST, etc.). The error syntax is captured in the http.Error calls. The bool returned is simply whether or not the queue was full. I'm open to a better way to capture this.

File: pkg/collector/pipeline.go [hunk 6]
```diff
@@ -47,37 +50,79 @@ func (c nowClock) Now() time.Time {
...
+	return setupPipeline(context.Background(), nil, bufferSize, numWorkers)
+}
+
+const defaultBufferSize = 1000
+const defaultNumWorkers = 10
+
+// NewTestPipeline creates a new Pipeline with a specified clock.
+// This should only be used for testing.
+func NewTestPipeline(clock Clock) *Pipeline {
+	return setupPipeline(context.Background(), clock, defaultBufferSize, defaultNumWorkers)
```
Comment (sburnett) [inline pkg/collector/pipeline.go]: nit: I would make this call NewTestPipeline with defaultBufferSize as an argument

File: pkg/collector/pipeline.go [hunk 7]
```diff
@@ -117,3 +165,11 @@ func (p *Pipeline) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	ctx := r.Context()
 	p.ProcessReports(ctx, w, r)
 }
+
+// Close stops the processing, such that anything in the queue
```
Comment (sburnett) [inline pkg/collector/pipeline.go]: I'd reword this to make it clear that it's the responsibility of the caller to make sure no one else tries to add more reports after calling this method. If I'm reading correctly, the program will panic if that happens (by sending to a closed channel).

File: pkg/collector/pipeline.go (hunk 1 above)
Comment (cstahl1) [reply pkg/collector/pipeline.go]: I fixed this to return an error, and added a special error for when it is dropped so that the caller can determine if it was dropped, successful, or something else.

File: pkg/collector/pipeline.go (hunk 7 above)
Comment (cstahl1) [reply pkg/collector/pipeline.go]: Done

File: pkg/collector/pipeline.go (hunk 6 above)
Comment (cstahl1) [reply pkg/collector/pipeline.go]: Done
